File: BEMLIB_22.10/01_num_meth/05_eigen/leverrier.f
      subroutine leverrier (N,A,c)

c======================================
c FDLIB
c Copyright by C. Pozrikidis, 1999
c All rights reserved
c
c This program is to be used only under the
c stipulations of the licensing agreement
c======================================

c------------------------------------------------
c This program accompanies the book:
c          C. Pozrikidis
c Numerical Computation in Science and Engineering
c        Oxford University Press
c------------------------------------------------

c-----------------------------------------
c  Coefficients of the 
c  characteristic polynomial of an nxn matrix 
c  by Leverrier' method
c
c  Algorithm (5.4.3)
c
c  SYMBOLS:
c  --------
c       
c  N .... size (rows/columns) of matrix A
c  A .... square matrix
c  c .... vector of polynomial coefficients
c
c  B .... iterated matrix for computation of c
c  D .... iterated matrix for computation of c
c
c P_N = (-1)^N * (lambda^N + c_1 \lambda^(N-1) + ... + c_N
c
c-----------------------------------------

      Implicit Double Precision (a-h,o-z)

      Dimension A(10,10),B(10,10),D(10,10)
      Dimension c(0:10)

c-----------------------------
c initial elements of matrix B 
c-----------------------------

      Do i=1,N
        Do j=1,N
          B(i,j) = 0.0D0
        End Do
        B(i,i) = 1.0D0
      End Do  

c-----------
c initialize
c-----------

      c(0) = 1.0D0

c---------------------------------------
c outer loop for polynomial coefficients
c---------------------------------------

      Do i=1,N             !  begin outer loop
   
      c(i) = 0.0D0

c---
c update the matrix D
c---

        Do j=1,N
         Do k=1,N
          D(j,k) = 0.0D0
           Do l=1,N
            D(j,k) = D(j,k) + A(j,l)*B(l,k)
           End Do
         End Do
        End Do

c---
c compute c(i)
c---
 
        tr = 0.0D0
 
        Do j=1,N
         tr = tr+D(j,j)
        End Do

        c(i) = -tr/i

c---
c update the matrix B
c---

        Do j=1,N
          Do k=1,N
           B(j,k) = D(j,k)
          End Do
          B(j,j) = B(j,j) + c(i)
        End Do

      End Do !  end of outer loop

c----------------------------------------------
c update coefficients for leading factor (-1)^N
c----------------------------------------------

      fact = (-1.0D0)**N

      Do i=1,N
       c(i) = fact*c(i)
      End Do

c-----
c done
c-----

      Return
      End

File: BEMLIB_22.10/01_num_meth/05_eigen/leverrier.m
      function c = leverrier(N,A)

%======================================
% FDLIB
% Copyright by C. Pozrikidis, 1999
% All rights reserved
%
% This program is to be used only under the
% stipulations of the licensing agreement
%======================================

%------------------------------------------------
% This program accompanies the book:
%          C. Pozrikidis
% Numerical Computation in Science and Engineering
%        Oxford University Press
%------------------------------------------------

%-----------------------------------------
%  Coefficients of the 
%  characteristic polynomial of an nxn matrix 
%  by Leverrier' method
%
%  Algorithm (5.4.3)
%
%  SYMBOLS:
%  --------
%       
%  N .... size (rows/columns) of matrix A
%  A .... square matrix
%  c .... vector of polynomial coefficients
%
%  B .... iterated matrix for computation of c
%  D .... iterated matrix for computation of c
%
% P_N = (-1)^N * (lambda^N + c_1*lambda^(N-1)
%     + ... + c_N
%-----------------------------------------

%-----------------------------
% initial elements of matrix B
%-----------------------------

  for i=1:N
    for j=1:N
      B(i,j) = 0.0;
    end
    B(i,i) = 1.0;
  end

%---------------------------------------
% outer loop for polynomial coefficients
%---------------------------------------

 for i=1:N   
   
   c(i) = 0.0;

%---
% update the matrix D
%---

  for j=1:N
   for k=1:N
    D(j,k) = 0.0;
     for l=1:N
      D(j,k) = D(j,k) + A(j,l)*B(l,k);
     end
   end
  end

%---
% compute c(i)
%---
 
  trace = 0.0;
 
  for j=1:N
   trace = trace + D(j,j);
  end

  c(i) = -trace/i;

%---
% update the matrix B
%---

   for j=1:N
     for k=1:N
       B(j,k) = D(j,k);
     end
     B(j,j) = B(j,j) + c(i);
   end

  end    % of outer loop

%----------------------------------------------
% update coefficients for leading factor (-1)^N
%----------------------------------------------

  factor = (-1.0)^N;

  for i=1:N
     c(i) = factor*c(i);
  end

%-----
% done
%-----

  return
